File: src/Flagrum.Blender/core/graphics/models/gamemodelbone.py
from dataclasses import dataclass

from ...serialization.messagepack.writer import MessagePackWriter


@dataclass(init=False)
class GameModelBone:
    name: str
    lod_index: int
    unique_index: int

    def read(self, reader):
        self.name = reader.read()
        self.lod_index = reader.read()

        if reader.data_version >= 20220707:
            self.unique_index = reader.read()
        else:
            self.unique_index = self.lod_index >> 16

    def write(self, writer: MessagePackWriter):
        writer.write(self.name)
        writer.write(self.lod_index)

File: src/Flagrum.Core/Graphics/Models/GameModelBone.cs
using Flagrum.Core.Serialization.MessagePack;

namespace Flagrum.Core.Graphics.Models;

public class GameModelBone : IMessagePackItem
{
    public string Name { get; set; }
    public uint LodIndex { get; set; }
    public uint UniqueIndex { get; set; }

    public void Read(MessagePackReader reader)
    {
        Name = reader.Read<string>();
        LodIndex = reader.Read<uint>();

        if (reader.DataVersion >= 20220707)
        {
            UniqueIndex = reader.Read<uint>();
        }
        else
        {
            UniqueIndex = LodIndex >> 16;
        }
    }

    public void Write(MessagePackWriter writer)
    {
        writer.Write(Name);
        writer.Write(LodIndex);

        if (writer.DataVersion >= 20220707)
        {
            writer.Write(UniqueIndex);
        }
    }
}